### PR TITLE
TS migrate: lib/client-api + app/angular + app/react + dependency upgrades

### PR DESCRIPTION
**Pull request by ndelangen**
combines #7147 & #6570

### CODE REVIEW

File: addons/a11y/src/components/Tabs.tsx [hunk 1]
```diff
@@ -77,7 +77,7 @@ const Item = styled.button(
 
 const TabsWrapper = styled.div({});
 
-const List = styled.div(({ theme }) => ({
+const List = styled.div<{}>(({ theme }) => ({
```
Comment (shilman) [inline addons/a11y/src/components/Tabs.tsx]: What's the benefit of `style.div<{}>(...)`? It's a bit ugyly and it seems like there are other places in this PR where it's still written `style.div(...)` so I'm just wondering what's the rationale here?

File: addons/events/src/components/Event.tsx [hunk 2]
```diff
@@ -17,7 +17,7 @@ interface StyledTextareaProps {
 
 const StyledTextarea = styled(({ shown, failed, ...rest }: StyledTextareaProps) => (
   <Textarea {...rest} />
-))(
+))<{ shown: boolean; failed: boolean }>(
```
Comment (shilman) [inline addons/events/src/components/Event.tsx]: what does this do?

File: addons/a11y/src/components/Tabs.tsx (hunk 1 above)
Comment (ndelangen) [reply addons/a11y/src/components/Tabs.tsx]: I wish i didn't have to:
https://github.com/emotion-js/emotion/issues/1298

File: addons/events/src/components/Event.tsx (hunk 2 above)
Comment (ndelangen) [reply addons/events/src/components/Event.tsx]: Adds types to the function that reduces theme+props into a style-object

File: app/angular/src/server/options.ts [hunk 3]
```diff
@@ -1,4 +1,5 @@
-import packageJson from '../../package.json';
+// tslint:disable-next-line: no-var-requires
+const packageJson = require('../../package.json');
```
Comment (backbone87) [inline app/angular/src/server/options.ts]: why was this changed?

Comment (ndelangen) [reply app/angular/src/server/options.ts]: Because this is outside the typescript root folder,

import doesn't work then, require (for some reason) does.

Comment (backbone87) [reply app/angular/src/server/options.ts]: til, still seems like an unsatisfying solution, i guess adding the `app/angular` to the `rootDirs` option wont work?

Comment (ndelangen) [reply app/angular/src/server/options.ts]: no, that causes far more grief.

File: lib/addons/src/types.ts [hunk 4]
```diff
@@ -11,3 +13,83 @@ export type Types = types | string;
 export function isSupportedType(type: Types): boolean {
   return !!Object.values(types).find(typeVal => typeVal === type);
 }
+
+export interface Parameters {
+  fileName?: string;
+  options?: OptionsParameter;
+  [key: string]: any;
```
Comment (backbone87) [inline lib/addons/src/types.ts]: do we really want to allow this wildcard?
addons can argument this interface to add their own params.

the same question applies to `StoryContext`

File: lib/addons/src/typings.d.ts [hunk 5]
```diff
@@ -1,3 +1,20 @@
 declare module 'global';
-declare module '@storybook/client-logger';
-declare module '@storybook/addons';
\ No newline at end of file
+
+declare global {
+  interface NodeModule {
```
Comment (backbone87) [inline lib/addons/src/typings.d.ts]: i think some TS config is gone rogue here. `__WebpackModuleApi.Module` (from which `NodeModule` extends when using types `webpack-env`) already declares `hot`

see https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/webpack-env/index.d.ts

Comment (ndelangen) [reply lib/addons/src/typings.d.ts]: yeah that makes more sense to use indeed

File: lib/addons/src/typings.d.ts [hunk 6]
```diff
@@ -1,3 +1,20 @@
...
+            decline(): void;
+
+            dispose(callback: (data: any) => void): void;
+            addDisposeHandler(callback: (data: any) => void): void;
+
+            removeDisposeHandler(callback: (data: any) => void): void;
+            // ...
+        }
+    }
+}
```
Comment (backbone87) [inline lib/addons/src/typings.d.ts]: trainling newline missing

File: lib/addons/src/types.ts (hunk 4 above)
Comment (ndelangen) [reply lib/addons/src/types.ts]: Yeah these are extensible by design.

Comment (backbone87) [reply lib/addons/src/types.ts]: i think we missunderstand.

```ts
export interface Parameters {
  fileName?: string;
  options?: OptionsParameter;
}
```
is just as extensible when using interface augmentation, it just doesnt allow untyped params (because no index signature). see https://www.typescriptlang.org/docs/handbook/declaration-merging.html#merging-interfaces

**Comment by ndelangen** [comment]
I merged the PRs individually
